specutils/analysis/flux.py

```diff
@@ -99,7 +99,15 @@ def _compute_line_flux(spectrum, regions=None):
     # Average dispersion in the line region
     avg_dx = (np.abs(np.diff(calc_spectrum.spectral_axis))).quantity
 
-    line_flux = np.sum(calc_spectrum.flux[1:] * avg_dx)
+    # Account for the existence of a mask.
+    if hasattr(spectrum, 'mask') and spectrum.mask is not None:
+        # Cannot use unmasked values because of average dispersion.
+        # Masked values must enter sum calculation valued as zeros.
+        flux_ = np.where(calc_spectrum.mask, 0, calc_spectrum.flux)[1:]
```

Comment:
So the `.mask` attribute is a boolean array where `True` means it is masked.  Because of this, it is probably a little clearer and more efficient (by a factor of 2) to do something like

```
flux_ = calc_spectrum.flux[~calc_spectrum.mask][1:]
```

And of course instead of returning an array of the same size with zeros in the masked areas, it is returning only the non-masked array values, which will also speed up the summation:

```
In [2]: a = np.arange(15).astype(np.float)

In [3]: mask = a > 9

In [4]: a[~mask]
Out[4]: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [5]: np.where(mask, 0, a)
Out[5]: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9., 0., 0., 0., 0., 0.])

In [6]: %timeit a[~mask]
765 ns ± 8.23 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [7]: %timeit np.where(mask, 0, a)
1.6 µs ± 38.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

I see you've used the above method in other areas of the code.  👍

Reply:
While this is true, it can have adverse affects on calculations that require the length of the array to be consistent (e.g. means):

```python
In [10]: flux[~mask]
Out[10]:
array([0.44012434, 0.18503118, 0.75635908, 0.80137293, 0.88490978,
       0.37460318])

In [11]: np.where(mask, 0, flux)
Out[11]:
array([0.44012434, 0.18503118, 0.        , 0.        , 0.75635908,
       0.80137293, 0.        , 0.88490978, 0.        , 0.37460318])

In [12]: np.mean(flux[~mask])
Out[12]: 0.5737334148059349

In [13]: np.mean(np.where(mask, 0, flux))
Out[13]: 0.3442400488835609
```

The route implemented now is certainly safer and more future-proof (since the `_compute_line_flux` is used as an intermediary in a lot of analysis functions), but this _could_ be circumvented by just being Very Careful if performance really is a huge bottleneck.

Reply:
In your example above though, only one mean is correct.  Masking doesn't mean treat the flux values as though they were 0, it means ignore them in the calculation.  By replacing the masked indices with a value of zero and then taking the mean, you're not ignoring them, but taking the mean of lots of zeros along with the good values, which gives the incorrect result, likely too low if the good values are positive.

Agree that one has to be careful.

Generally, each analysis function needs to take the input mask and do what is correct for that function I suspect.  And this will vary by function, as some might also depend on the binning or array length.  Because of this, I don't think it is a good idea to carry around various private class attributes that already have the mask applied in a particular way to the array.  Better to apply the mask, function-by-function as needed.